File: exercism/clojure/log-levels/src/log_levels.clj
(ns log-levels
  (:require [clojure.string :as str]))

(defn message
  "Takes a string representing a log line
   and returns its message with whitespace trimmed."
  [s]
  (cond 
    (.startsWith s "[ERROR]:") (.strip (.substring s (count "[ERROR]:")))
    (.startsWith s "[WARNING]:") (.strip (.substring s (count "[WARNING]:")))
    (.startsWith s "[INFO]:") (.strip (.substring s (count "[INFO]:")))
    )
  )

(defn log-level
  "Takes a string representing a log line
   and returns its level in lower-case."
  [s]
    (cond
      (.startsWith s "[ERROR]:") "error"
      (.startsWith s "[WARNING]:") "warning"
      (.startsWith s "[INFO]:") "info")
  )

(defn reformat
  "Takes a string representing a log line and formats it
   with the message first and the log level in parentheses."
  [s]
  (str (message s) " (" (log-level s) ")")
  )

File: exercism/java/log-levels/src/main/java/LogLevels.java
public class LogLevels {
    
    private static String errorStart = "[ERROR]: ";
    private static String warningStart = "[WARNING]: ";
    private static String infoStart = "[INFO]: ";
    public static String message(String logLine) {
        if(logLine.startsWith(errorStart)){
            return logLine.substring(errorStart.length()).strip();
        }
        else if(logLine.startsWith(warningStart)){
            return logLine.substring(warningStart.length()).strip();
        }
        else if(logLine.startsWith(infoStart)){
            return logLine.substring(infoStart.length()).strip();
        }
        return logLine;
    }

    public static String logLevel(String logLine) {
        if(logLine.startsWith(errorStart)){
            return "error";
        }
        else if(logLine.startsWith(warningStart)){
            return "warning";
        }
        else if(logLine.startsWith(infoStart)){
            return "info";
        }
        return null;
    }

    public static String reformat(String logLine) {
        return message(logLine)+" (%s)".formatted(logLevel(logLine));
    }
}
